File: Algorithms on Graphs/Assignment4/negative_cycle.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Graph {
private:
    vector<vector<int> > adj;
    vector<vector<int> > cost;
    vector<int> distance;
    vector<int> prev;
    
    int find_index(vector<int>& vec, int i) {
        vector<int>::iterator ind_itr = find(vec.begin(), vec.end(), i);
        int ind = (int)(ind_itr - vec.begin());
        return ind;
    }
    
    void relax(int u, int v) {
        int ind = find_index(adj[u], v);
        if (distance[v] > distance[u] + cost[u][ind]) {
            distance[v] = distance[u] + cost[u][ind];
            prev[v] = u;
        }
    }
    
    int bellman_ford(int s) {
        bool flag = 0;
        distance[s] = 0;
        for (int j = 0; j < adj.size() - 1; j++) {
            for (int i = 0; i < adj.size(); i++) {
                for (int v: adj[i]) {
                    relax(i, v);
                }
            }
        }
        for (int i = 0; i < adj.size(); i++) {
            for (int v:adj[i]) {
                int ind = find_index(adj[i], v);
                if (distance[v] > distance[i] + cost[i][ind]) {
                    distance[v] = distance[i] + cost[i][ind];
                    prev[v] = i;
                    flag = 1;
                }
            }
        }
        return flag;
    }
    
public:
    Graph(vector<vector<int> > &adj_, vector<vector<int> > &cost_):
    adj(adj_),
    cost(cost_),
    prev(vector<int>(adj_.size(), -1)),
    distance(vector<int>(adj_.size(), numeric_limits<int>::infinity()))
    {}
    
    int negative_cycle() {
        int flag = bellman_ford(0);
        return flag;
    }
};

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    vector<vector<int> > cost(n, vector<int>());
    for (int i = 0; i < m; i++) {
        int x, y, w;
        cin >> x >> y >> w;
        adj[x - 1].push_back(y - 1);
        cost[x - 1].push_back(w);
    }
    Graph g = Graph(adj, cost);
    cout << g.negative_cycle();
}

File: Algorithms on Graphs/Assignment4/negative_cycle.py
#Uses python3

import sys

class Graph:
    def __init__(self, adj_, cost_):
        self.adj = adj_
        self.cost = cost_
        self.distance = [float('inf') for _ in range(len(adj_))]
        self.prev = [-1 for _ in range(len(adj_))]

    def relax(self, u, v):
        if self.distance[v] > self.distance[u] + self.cost[u][self.adj[u].index(v)]:
            self.distance[v] = self.distance[u] + self.cost[u][self.adj[u].index(v)]
            self.prev[v] = u



    def Bellman_Ford(self, s):
        self.distance[s] = 0
        self.prev[s] = None

        for i in range(len(self.adj) - 1):
            for i in range(len(self.adj)):
                for v in self.adj[i]:
                    self.relax(i, v)

        for i in range(len(self.adj)):
            for v in self.adj[i]:
                if self.distance[v] > self.distance[i] + self.cost[i][self.adj[i].index(v)]:
                    self.distance[v] = self.distance[i] + self.cost[i][self.adj[i].index(v)]
                    self.prev[v] = i
                    return 1
        return 0

    def negative_cycle(self):
        flag = self.Bellman_Ford(0)
        return flag

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    #data = [4, 4, 1, 2, -5, 4, 1, 2, 2, 3, 2, 3, 1, 1]
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(zip(data[0:(3 * m):3], data[1:(3 * m):3]), data[2:(3 * m):3]))
    data = data[3 * m:]
    adj = [[] for _ in range(n)]
    cost = [[] for _ in range(n)]
    for ((a, b), w) in edges:
        adj[a - 1].append(b - 1)
        cost[a - 1].append(w)
    g = Graph(adj, cost)
    print(g.negative_cycle())
